src/libstore/s3.cc

```diff
@@ -7,11 +8,183 @@
+        }
+    });
+}
+
+// AwsCredentialProvider implementation
+
+std::unique_ptr<AwsCredentialProvider> AwsCredentialProvider::createDefault()
+{
+    initAwsCrt();
+
+    if (!crtInitialized) {
+        throw AwsAuthError("AWS CRT not initialized, cannot create credential provider");
+    }
+
+    try {
+        Aws::Crt::Auth::CredentialsProviderChainDefaultConfig config;
+        config.Bootstrap = Aws::Crt::ApiHandle::GetOrCreateStaticDefaultClientBootstrap();
+
+        auto provider = Aws::Crt::Auth::CredentialsProvider::CreateCredentialsProviderChainDefault(config);
+        if (!provider) {
+            throw AwsAuthError("Failed to create default AWS credentials provider");
+        }
+
+        return std::make_unique<AwsCredentialProvider>(provider);
+    } catch (const AwsAuthError &) {
+        throw;
+    } catch (const std::exception & e) {
+        throw AwsAuthError("Exception creating AWS credential provider: %s", e.what());
+    } catch (...) {
+        throw AwsAuthError("Unknown exception creating AWS credential provider");
+    }
+}
+
+std::unique_ptr<AwsCredentialProvider> AwsCredentialProvider::createProfile(const std::string & profile)
+{
+    initAwsCrt();
+
+    if (!crtInitialized) {
+        throw AwsAuthError("AWS CRT not initialized, cannot create credential provider");
+    }
+
+    if (profile.empty()) {
+        return createDefault();
+    }
+
+    try {
+        Aws::Crt::Auth::CredentialsProviderProfileConfig config;
+        config.Bootstrap = Aws::Crt::ApiHandle::GetOrCreateStaticDefaultClientBootstrap();
+        config.ProfileNameOverride = Aws::Crt::ByteCursorFromCString(profile.c_str());
+
+        auto provider = Aws::Crt::Auth::CredentialsProvider::CreateCredentialsProviderProfile(config);
+        if (!provider) {
+            throw AwsAuthError("Failed to create AWS credentials provider for profile '%s'", profile);
+        }
+
+        return std::make_unique<AwsCredentialProvider>(provider);
+    } catch (const AwsAuthError &) {
+        throw;
+    } catch (const std::exception & e) {
+        throw AwsAuthError("Exception creating AWS credential provider for profile '%s': %s", profile, e.what());
+    } catch (...) {
+        throw AwsAuthError("Unknown exception creating AWS credential provider for profile '%s'", profile);
+    }
+}
+
+AwsCredentialProvider::AwsCredentialProvider(std::shared_ptr<Aws::Crt::Auth::ICredentialsProvider> provider)
+    : provider(provider)
+{
+}
+
+AwsCredentialProvider::~AwsCredentialProvider() = default;
+
+AwsCredentials AwsCredentialProvider::getCredentials()
+{
+    if (!provider || !provider->IsValid()) {
+        throw AwsAuthError("AWS credential provider is invalid");
+    }
+
+    std::mutex mutex;
+    std::condition_variable cv;
+    std::optional<AwsCredentials> result;
+    int resolvedErrorCode = 0;
+    bool resolved = false;
+
+    provider->GetCredentials([&](std::shared_ptr<Aws::Crt::Auth::Credentials> credentials, int errorCode) {
```

Comment:
Is `GetCredentials` asynchronous? What happens if the callback isn't called? Will `cv.wait()` hang forever?